### PR TITLE
Migrate stream_ordering to a bigint

### LINKED ISSUE #8255: events table uses a 32-bit integer for stream_ordering
2 billion events isn't actually that many; current estimate for exhaustion on matrix.org is Q3 2021, but that's an optimistic estimate as the rate of events is increasing.

nb there are other tables that use a standard int too.

### PR DESCRIPTION
**Pull request by richvdh**
A set of background updates (mostly) to replace `events.stream_ordering` with a `BIGINT`.

Fixes at least part of #8255.

The general approach here is:

1. Add a new `BIGINT` column called `stream_ordering2`.
2. Add a Postgres rule to populate that column with the same value as `stream_ordering` for any new rows.
3. Run a bg update to back-populate the column for all *existing* rows.
4. Add an index on the new column.
5. In a single transaction:
    * Double-check that we didn't miss any rows with a final `UPDATE`
    * Drop the rule
    * Drop the old column
    * Rename the new column into place

I still need to check that we don't have any ints in other tables.

### CODE REVIEW

**Comment by callahad** [comment]
Do we also need to worry about `last_stream_ordering` on the `pushers` table?

**Comment by richvdh** [comment]
> Do we also need to worry about `last_stream_ordering` on the `pushers` table?

Probably! I have yet to work through other tables to see what other problems we have. I'm assuming they are all going to be easier than `events`.

File: synapse/storage/databases/main/events_bg_updates.py [hunk 1]
```diff
@@ -1009,3 +1043,75 @@ def purged_chain_cover_txn(txn) -> int:
...
+
+    async def _background_populate_stream_ordering2(
+        self, progress: JsonDict, batch_size: int
+    ) -> int:
+        """Populate events.stream_ordering2, then replace stream_ordering
+
+        This is to deal with the fact that stream_ordering was initially created as a
+        32-bit integer field.
+        """
+        batch_size = max(batch_size, 1)
```
Comment (erikjohnston) [inline synapse/storage/databases/main/events_bg_updates.py:1055]: Why is this an issue here? I'd expect setting batch_size to zero/negative anywhere is going to cause issues?

**Review by erikjohnston — COMMENTED** [review]
At a *very* cursory glance this looks sane

**Comment by anoadragon453** [comment]
> Add an index on the new column.

Is this expected to be a long-running task?

**Comment by richvdh** [comment]
> > Add an index on the new column.
> 
> Is this expected to be a long-running task?

well yes. But the creation runs concurrently so shouldn't block other operations.

File: synapse/storage/databases/main/events_bg_updates.py (hunk 1 above)
Comment (richvdh) [reply synapse/storage/databases/main/events_bg_updates.py:1055]: true. It's just defensive here.

**Review by babolivier — APPROVED** [review]

**Pull request merged**